arch/x86/core/pcie.c

```diff
@@ -58,99 +59,137 @@ static void pcie_mm_init(void)
+
+#define PCIE_IO_CONF(rw, size, type)                   \
+static inline void pcie_io_conf_##rw##_##size          \
+	(pcie_bdf_t bdf, unsigned int addr, type data) \
+{                                                      \
+	k_spinlock_key_t k;                            \
+\
+	bdf &= PCIE_X86_CAP_BDF_MASK;                  \
+	bdf |= PCIE_X86_CAP_EN;                        \
+	bdf |= (addr & PCIE_X86_CAP_WORD_MASK);        \
+\
+	k = k_spin_lock(&pcie_io_conf_spinlock);       \
+	sys_out32(bdf, PCIE_X86_CAP);                  \
+\
+	PCIE_IO_CONF_##rw(size, data);                 \
+\
+	sys_out32(0U, PCIE_X86_CAP);                   \
+	k_spin_unlock(&pcie_io_conf_spinlock, k);      \
+}
 
-	bdf &= PCIE_X86_CAP_BDF_MASK;
-	bdf |= PCIE_X86_CAP_EN;
-	bdf |= (reg & PCIE_X86_CAP_WORD_MASK) << PCIE_X86_CAP_WORD_SHIFT;
+PCIE_IO_CONF(read, 8, uint8_t *)
+PCIE_IO_CONF(read, 16, uint16_t *)
+PCIE_IO_CONF(read, 32, uint32_t *)
 
-	k = k_spin_lock(&lock);
-	sys_out32(bdf, PCIE_X86_CAP);
+PCIE_IO_CONF(write, 8, uint8_t)
+PCIE_IO_CONF(write, 16, uint16_t)
+PCIE_IO_CONF(write, 32, uint32_t)
 
-	if (write) {
-		sys_out32(*data, PCIE_X86_CDP);
-	} else {
-		*data = sys_in32(PCIE_X86_CDP);
-	}
-
-	sys_out32(0U, PCIE_X86_CAP);
-	k_spin_unlock(&lock, k);
-}
-
-static inline void pcie_conf(pcie_bdf_t bdf, unsigned int reg,
-			     bool write, uint32_t *data)
+/* these functions are explained in include/drivers/pcie/pcie.h */
 
-{
 #ifdef CONFIG_PCIE_MMIO_CFG
-	if (bus_segs[0].mmio == NULL) {
-		pcie_mm_init();
-	}
-
-	if (do_pcie_mmio_cfg) {
-		pcie_mm_conf(bdf, reg, write, data);
-	} else
+#define IF_CONFIG_PCIE_MMIO_CFG(...) __VA_ARGS__
+#else
+#define IF_CONFIG_PCIE_MMIO_CFG(...)
 #endif
-	{
-		pcie_io_conf(bdf, reg, write, data);
-	}
+
+#define PCIE_CONF_VALIDATE_write(size, data) 0
+#define PCIE_CONF_VALIDATE_read(size, data) \
+	((*data == (uint##size##_t)(~0x0)) ? -ENXIO : 0)
+
+#define PCIE_CONF(rw, size, type)                                         \
+int pcie_conf_##rw##_##size(pcie_bdf_t bdf, unsigned int addr, type data) \
+{                                                                         \
+	/* validate alignment */                                          \
+	if (addr & ((size / 8) - 1)) {                                    \
+		return -EINVAL;                                           \
+	}                                                                 \
+IF_CONFIG_PCIE_MMIO_CFG(                                                  \
+	if (bus_segs[0].mmio == NULL) {                                   \
+		pcie_mm_init();                                           \
+	}                                                                 \
+	if (do_pcie_mmio_cfg) {                                           \
+		pcie_mm_conf_##rw##_##size(bdf, addr, data);              \
+	} else                                                            \
+		)                                                         \
```

Comment:
Might want to align this with the `IF_CONFIG_PCIE_MMIO_CFG` above.

Reply:
This is how I originally wrote this, but checkpatch does not allow indent 0 after the else keyword. Is there a way to tell checkpatch to ignore this?

Reply:
Hm it does not complain about it this time. I updated the alignment